californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java

```diff
@@ -1016,6 +1017,7 @@ public ExchangeCallback(final Exchange exchange, final Message message) {
 
 		@Override
 		public void onContextEstablished(EndpointContext context) {
+			exchange.setSendTimestamp(ClockUtil.nanoRealtime());
```

Comment:
Any reason to set this onContextEstablished instead of onSent ?
(I didn't remember how this behave in case of blockwise transfer :thinking:)

(should we also setTimestamp directly on message in case exchange is null (see line 776))

Reply:
`onContextEstablished` is the last, before the message is sent. If a request is received before that, then it could be ignored, because it's ensured, that the response was not sent before.

`onSent` is the first, after the message is sent. If a request is received before that, then it's not sure, if the response was not sent before.

The difference may be very small ...

Reply:
> (should we also setTimestamp directly on message in case exchange is null (see line 776))

I know, that the MID and token is set to the Request (even from an other thread). I don't like to alter too much in parallel. Therefore I used a timestamp within the exchange.